KAFKA-6383: complete shutdown for CREATED StreamThreads

rodesai opened the pull request:
When transitioning StreamThread from CREATED to PENDING_SHUTDOWN
free up resources from the caller, rather than the stream thread,
since in this case the stream thread was never actually started.

In KakfaStreams.close, shut down the streams threads from the
close thread. StreamThread.shutdown may now block, so call this
from the close thread so that the timeout is honored.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

rodesai committed:
KAFKA-6383: complete shutdown for CREATED StreamThreads

When transitioning StreamThread from CREATED to PENDING_SHUTDOWN
free up resources from the caller, rather than the stream thread,
since in this case the stream thread was never actually started.

In KakfaStreams.close, shut down the streams threads from the
close thread. StreamThread.shutdown may now block, so call this
from the close thread so that the timeout is honored.

rodesai commented:
@mjsax @guozhangwang

rodesai committed:
KAFKA-6383: Fix shutdown unit tests

Have the shutdown unit test shut down the thread from the
created->running state change callback, called from run().
This way we get past the state check in run().

Add a unit test to verify that the task manager is shut down
even if start() is never called.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -714,7 +720,10 @@ public StreamThread(final Time time,
     public void run() {
         log.info("Starting");
-        setState(State.RUNNING);
+        if (!setState(State.RUNNING)) {
```
Could this ever happen? I.e. instead of checking the lower-level tread state here, we already check the higher-level instance state that in `KafkaStreams#start` if the instance state is not `CREATED` we will fail and throw?

rodesai replied:
If the client does:
```
KafkaStreams ks = ...
ks.start();
ks.close();
```
Isn't it possible that the StreamThread.shutdown() gets called, sees that the thread is still in CREATED state and starts shutting the stream thread down before the thread itself gets scheduled or gets far enough into run() ?

guozhangwang replied:
Ack, makes sense.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -796,23 +796,23 @@ public synchronized boolean close(final long timeout, final TimeUnit timeUnit) {
             final Thread shutdownThread = new Thread(new Runnable() {
                 @Override
                 public void run() {
+                    // notify all the threads to stop; avoid deadlocks by stopping any
```
Is it necessary to move this logic within the close thread? Maybe I'm still missing something..

rodesai replied:
My thinking was that since ThreadState.shutdown() can now call ThreadState.completeShutdown(), its now possible that ThreadState.shutdown() will block on the producer exiting (or something else exiting... it seems to be closing a lot of stuff). If this wait is indeterminately long, then we need to call shutdown() from the close thread to make sure we honor the timeout passed in to close().

guozhangwang replied:
Ack, that makes sense.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -209,6 +211,10 @@ boolean setState(final State newState) {
         }
 
+        if (oldState == State.CREATED && newState == State.PENDING_SHUTDOWN) {
+            completeShutdown(true);
```
I'm wondering if we can avoid injecting the thread logic in the `setState` function. One idea is to not allow the transition from `CREATED` to `PENDING_SHUTDOWN`, and hence the `setState(PENDING_SHUTDOWN)` in `KafkaStreams#shutdown()` will return false (note all other states except `DEAD` can transit to `PENDING_SHUTDOWN`), and then we can depend on the return value i.e. in `shutdown()`:

```
if (!setState(PENDING_SHUTDOWN) && state() == CREATED)) {
    completeShutdown(true);
}
```

rodesai replied:
If we did that we could still race with with the StreamThread itself and do stuff in run() while the caller of shutdown() is in completeShutdown(). I do agree that putting this logic in setState is ugly. Another option would be to have a casState (compare and swap state) function so the caller can be sure which state transition its making.

mjsax replied:
But couldn't we have a lock on the state within `completeShutdown()` ? Thus, if we are not running yet, and enter `completeShutdown()` we also don't allow to transit into RUNNING -- and if we did leave CREATED, we "cancel" `completeShutdown()` ?

guozhangwang reviewed (commented):
@rodesai thanks for the patch, left some comments.

guozhangwang commented:
I think your suggestion of `casState` is more promising. More specifically, we can actually modify the `setState()` return type as the old state, and if the setState fails return `null`. Here is my reasoning:

There are two calling sequences that matters.

1) `StreamThread#run()` is never called since `KafkaStreams#start()` is not called. In this case we can let the caller thread of `KafkaStreams#shutdown()` to help trigger starting the stream threads after setting their state to `PENDING_SHUTDOWN`, we know that no other callers would call `thread#start()` again since it is guarded by the instance state.

2) `StreamThread#run()` is eventually called but the thread state is no longer `CREATED`.  In this case the thread should just proceed to call `completeShutdown`.

So it is still doable to only let the stream threads themselves to do the cleanup, to keep logic simpler to reason. To sum:

1) in `KafkaStreams#close()` we call `setState(PENDING_SHUTDOWN)` and then if it returns `CREATED` go ahead to call `thread.start()` after `thread.shutdown()`; if it returns others just call `thread.start()` (as is today); if it returns null it means the old state is `NOT_RUNNING` and it should not happen and hence we throw and fail.

2) in `KafkaStreams#start()` we call `setState(RUNNING)`, and if it returns null it means either `PENDING_SHUTDOWN` or `NOT_RUNNING`, and hence we do not need to proceed further, and just return.

3) in `StreamThread#run()` function if `setState(RUNNING)` returns null we can just proceed to `completeShutdown()`.

WDYT?

rodesai commented:
That works. We can move the calls to shutdown back out of the clean up thread too.

That said, do you think it would be cleaner to move the
```
if (setState(PENDING_SHUTDOWN) == CREATED) {
    thread.start();
}
```
logic into StreamThread.shutdown(). This gives StreamThread cleaner semantics. As long as shutdown() gets called on it, everything gets cleaned up. The client doesn't have to worry about calling start() in that case.

mjsax reviewed (commented):
We should add a test that calls `close()` without previous `start()` and pass in a mocked client to verify that those get closed (`KafkaStreams` allows to pass in a client supplier in it's constructor).

guozhangwang commented:
> logic into StreamThread.shutdown(). This gives StreamThread cleaner semantics. As long as shutdown() gets called on it, everything gets cleaned up. The client doesn't have to worry about calling start() in that case.

Sounds good to me.

rodesai committed:
Address some review feedback

Have StreamThread.setState return the old state. If the old state is
CREATED in StreamThread.shutdown() then start the thread so that it
can free the resources owned by the StreamThread.

Add a KafkaStreams test to verify that the producer gets closed even
if KafkaStreams was not started

rodesai committed:
Move comment back to right place

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -174,22 +174,24 @@ public State state() {
      */
-    boolean setState(final State newState) {
-        final State oldState = state;
+    State setState(final State newState) {
```
Nit: can we update JavaDoc with `@return`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -274,7 +274,11 @@ void shutdown(final boolean clean) {
-        restoreConsumer.unsubscribe();
+        try {
+            restoreConsumer.unsubscribe();
+        } catch (final RuntimeException fatalException) {
```
Why this change? AFAIK, `unsubscribe()` will never throw.

rodesai replied:
It throws in the test since the consumer was never opened.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -113,6 +115,28 @@ public void testStateCloseAfterCreate() {
     }
 
+    @Test
+    public void testStreamsCloseWithoutStart() throws Exception {
```
nit: rename `shouldCleanupResourcesOnCloseWithoutPreviousStart()`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -455,6 +455,35 @@ public void shouldShutdownTaskManagerOnClose() throws InterruptedException {
+        EasyMock.expectLastCall();
+        EasyMock.replay(taskManager, consumer);
+
+        StreamThread.StreamsMetricsThreadImpl streamsMetrics = new StreamThread.StreamsMetricsThreadImpl(metrics, "", "", Collections.<String, String>emptyMap());
```
nit: add `final` and line too long

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -455,6 +455,35 @@ public void shouldShutdownTaskManagerOnClose() throws InterruptedException {
+        EasyMock.replay(taskManager, consumer);
+
+        StreamThread.StreamsMetricsThreadImpl streamsMetrics = new StreamThread.StreamsMetricsThreadImpl(metrics, "", "", Collections.<String, String>emptyMap());
+        final StreamThread thread = new StreamThread(mockTime,
```
nit: fix indention -- either move `mockTime` one line down or indent other parameters to align with `mockTime`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -455,6 +455,35 @@ public void shouldShutdownTaskManagerOnClose() throws InterruptedException {
+        thread.shutdown();
+        try {
+            thread.join(1000);
+        } catch (InterruptedException e) {
```
nit: add `final`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -1088,7 +1094,11 @@ private long computeLatency() {
+        State oldState = setState(State.PENDING_SHUTDOWN);
+        if (oldState == State.CREATED) {
+            // Start so that we shutdown on the thread
+            this.start();
```
Why don't we call `completeShutdown(true)` directly instead via the redirect to `run()` ?

rodesai replied:
It seemed simpler to have all the shutdown stuff handled in the streams thread. I don't have a strong preference either way though.

guozhangwang replied:
This lgtm.

rodesai committed:
Fix some review comments

guozhangwang commented:
LGTM. Merged to trunk.
